include/circt/Dialect/ESI/ESIOps.td

```diff
@@ -82,17 +82,43 @@ def CosimEndpoint : ESI_Physical_Op<"cosim", []> {
     supported by the specific protocol.
   }];
 
-  let arguments = (ins I1:$clk, I1:$rstn, ChannelType:$input,
+  let arguments = (ins I1:$clk, I1:$rstn, ChannelType:$send,
     I64Attr:$endpointID);
-  let results = (outs ChannelType:$output);
+  let results = (outs ChannelType:$recv);
 
   let assemblyFormat = [{
-    $clk `,` $rstn `,` $input `,` $endpointID attr-dict
-      `:` type($input) `->` type($output)
+    $clk `,` $rstn `,` $send `,` $endpointID attr-dict
+      `:` type($send) `->` type($recv)
   }];
 
   let extraClassDeclaration = [{
-    uint64_t getInputTypeID() { return getCapnpTypeID(input().getType()); }
-    uint64_t getOutputTypeID() { return getCapnpTypeID(output().getType()); }
+    uint64_t getSendTypeID() { return getCapnpTypeID(send().getType()); }
+    uint64_t getRecvTypeID() { return getCapnpTypeID(recv().getType()); }
+  }];
+}
+
+def RtlBitArray : Type<CPred<"$_self.isa<::circt::rtl::ArrayType>()"
+ " && $_self.cast<::circt::rtl::ArrayType>().getElementType() =="
+ "   ::mlir::IntegerType::get($_self.getContext(), 1)">, "an RTL bit array">;
+
+def CapnpIngest : ESI_Physical_Op<"ingest.capnp", [NoSideEffect]> {
+  let summary = "Translate bits in Cap'nProto messages to RTL typed data";
+
+  let arguments = (ins RtlBitArray:$capnpBits);
+  let results = (outs AnyType:$decodedData);
+
+  let assemblyFormat = [{
+    $capnpBits attr-dict `:` type($capnpBits) `->` type($decodedData)
+  }];
+}
+
+def CapnpEgest : ESI_Physical_Op<"egest.capnp", [NoSideEffect]> {
```

Comment:
Minor comment, but I think it would help for me at least: instead of naming it ingest/egest, what about decode/encode? It's a little hard for me to remember what an `esi.egest.capnp` does versus `esi.encode.capnp`. This may be a personal, problem, I hadn't even heard the word "egest" before.

Reply:
I had to look up the antonym to 'ingest' myself. Good suggestion.